testcases/kernel/syscalls/inotify/inotify07.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2017 CTERA Networks.  All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it would be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * Further, this software is distributed without any warranty that it is
+ * free of the rightful claim of any third person regarding infringement
+ * or the like.  Any license provided herein, whether implied or
+ * otherwise, applies only to this software file.  Patent licenses, if
+ * any, provided herein do not apply to combinations of this program with
+ * other software, or any other product whatsoever.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Started by Amir Goldstein <amir73il@gmail.com>
+ *
+ * DESCRIPTION
+ *     Check that inotify work for a directory after drop caches
+ *
+ * ALGORITHM
+ *     Add watch on a directory and drop caches dentry and inode cache.
+ *     inotify pins the directory inode in cache, but not the dentry.
+ *     Execute operations on directory and child and expect events to be
+ *     reported on directory watch. This will fail if file system does
+ *     not obtain the pinned inode to the new allocated dentry after drop
+ *     caches.
+ */
+
+#include "config.h"
+
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <limits.h>
+#include "tst_test.h"
+#include "inotify.h"
+
+#if defined(HAVE_SYS_INOTIFY_H)
+#include <sys/inotify.h>
+
+#define TST_TOTAL 4
+
+#define EVENT_MAX 1024
+/* size of the event structure, not counting name */
+#define EVENT_SIZE  (sizeof (struct inotify_event))
+/* reasonable guess as to size of 1024 events */
+#define EVENT_BUF_LEN        (EVENT_MAX * (EVENT_SIZE + 16))
+
+#define BUF_SIZE 256
+static int fd, fd_notify, reap_wd;
+static int wd;
+
+struct event_t {
+	char name[BUF_SIZE];
+	unsigned int mask;
+};
+
+#define DIR_NAME "test_dir"
+#define FILE_NAME1 "test_file1"
+#define FILE_NAME2 "test_file2"
+
+static struct event_t event_set[EVENT_MAX];
+
+static char event_buf[EVENT_BUF_LEN];
+
+static const char drop_caches_fname[] = "/proc/sys/vm/drop_caches";
+
+static void drop_caches(void)
+{
+	int ret;
+	FILE *f;
+
+	f = fopen(drop_caches_fname, "w");
+	if (f) {
+		/* Drop inode and dentry caches */
+		ret = fprintf(f, "2");
+		fclose(f);
+		if (ret < 1)
+			tst_brk(TBROK, "Failed to drop caches");
+	} else {
+		tst_brk(TBROK, "Failed to open drop_caches");
+	}
+}
+
```

Comment:
We do have SAFE_FILE_PRINTF() exactly for this purpose, the whole function can be replaced with:
SAFE_FILE_PRINTF("proc/sys/vm/drop_caches", "2");